### PR TITLE
 feat: Implement validate_coinbase function #14 

### PR DESCRIPTION
**Pull request by manlikeHB**
- [ ] issue #14
- [ ] follows contribution [guide](https://github.com/keep-starknet-strange/raito/blob/main/CONTRIBUTING.md)
- [ ] code change includes tests

 Implement validate_coinbase function

### CODE REVIEW

File: src/validation.cairo [hunk 1]
```diff
@@ -193,8 +196,38 @@ fn fee_and_merkle_root(block: @Block) -> Result<(u256, u256), ByteArray> {
     Result::Ok((total_fee, merkle_root(ref txids)))
 }
 
-fn validate_coinbase(block: @Block, total_fees: u256) -> Result<(), ByteArray> {
-    //TODO implement
+fn validate_coinbase(self: @ChainState, block: @Block, total_fees: u256) -> Result<(), ByteArray> {
```
Comment (manlikeHB) [inline src/validation.cairo:199]: @maciejka, I added 'self: ChainState' as a parameter to get access to the block_height to be able to computing block reward.

File: src/validation.cairo [hunk 2]
```diff
@@ -193,8 +196,38 @@ fn fee_and_merkle_root(block: @Block) -> Result<(u256, u256), ByteArray> {
...
+    // Validate the outputs' amounts
+    // Sum up the total amount of all outputs
+    // and also add the outputs to the UtreexoSet.
+    let mut total_output_amount = 0;
+
+    for txs in *block
+        .txs {
+            for outputs in *txs.outputs {
+                total_output_amount + *outputs.value;
+                //TODO add outputs to UtreexoSet
```
Comment (manlikeHB) [inline src/validation.cairo:221]: I'm not really sure how to go about adding the outputs to UtreexoSet.

File: src/validation.cairo [hunk 3]
```diff
@@ -193,8 +196,41 @@ fn fee_and_merkle_root(block: @Block) -> Result<(u256, u256), ByteArray> {
...
 }
 
-fn validate_coinbase(block: @Block, total_fees: u256) -> Result<(), ByteArray> {
-    //TODO implement
+fn validate_coinbase(self: @ChainState, block: @Block, total_fees: u256) -> Result<(), ByteArray> {
+    let tx_context = block.txs[0];
+
+    // Validate the coinbase input
+    // Ensure there is exactly one coinbase input
+    assert(tx_context.inputs.clone().len() == 1, 'Input count should be 1');
```
Comment (maciejka) [inline src/validation.cairo:204]: No need to clone here. You should rather use `(*tx_context.inputs).len()`.

File: src/validation.cairo [hunk 4]
```diff
@@ -193,8 +196,41 @@ fn fee_and_merkle_root(block: @Block) -> Result<(u256, u256), ByteArray> {
     Result::Ok((total_fee, merkle_root(ref txids)))
 }
 
-fn validate_coinbase(block: @Block, total_fees: u256) -> Result<(), ByteArray> {
-    //TODO implement
+fn validate_coinbase(self: @ChainState, block: @Block, total_fees: u256) -> Result<(), ByteArray> {
+    let tx_context = block.txs[0];
```
Comment (maciejka) [inline src/validation.cairo:200]: Why not call it just `tx`? It makes reading easier.

File: src/validation.cairo [hunk 5]
```diff
@@ -193,8 +196,41 @@ fn fee_and_merkle_root(block: @Block) -> Result<(u256, u256), ByteArray> {
...
+    assert(*tx_context.inputs[0].previous_output.txid == 0, 'txid should be 0');
+
+    // Validate the outputs' amounts
+    // Sum up the total amount of all outputs
+    // and also add the outputs to the UtreexoSet.
+    let mut total_output_amount: u64 = 0;
+
+    for txs in *block
+        .txs {
+            for outputs in *txs
```
Comment (maciejka) [inline src/validation.cairo:219]: `output` not `outputs`

File: src/validation.cairo [hunk 6]
```diff
@@ -193,8 +196,41 @@ fn fee_and_merkle_root(block: @Block) -> Result<(u256, u256), ByteArray> {
...
+                .outputs {
+                    total_output_amount + (*outputs.value).try_into().unwrap();
```
Comment (maciejka) [inline src/validation.cairo:221]: I think it is an error in the `TxOut.value` type, it should be `i64` not `u64`. Please fix it there, and remove `try_into` here.

File: src/validation.cairo (hunk 1 above)
Comment (maciejka) [reply src/validation.cairo:199]: Just add `block_height` as parameter. It will be easier to write tests.

**Review by maciejka — CHANGES_REQUESTED** [review]

File: src/validation.cairo [hunk 7]
```diff
@@ -350,4 +386,182 @@ mod tests {
...
+                }
+            ]
+                .span()
+        };
+        let total_fees = 5000000000_u64;
+        let block_height = 1;
+
+        validate_coinbase(@block, total_fees.into(), block_height);
+    }
+    // #[test]
```
Comment (manlikeHB) [inline src/validation.cairo:525]: I've been trying to make this test pass but I'm not sure what I'm missing @maciejka @m-kus,

File: src/validation.cairo [hunk 8]
```diff
@@ -350,4 +386,182 @@ mod tests {
...
+// #[should_panic(expected: ('total output > block rwd + fees',))]
+// fn test_validate_coinbase_outputs_amount() {
```
Comment (manlikeHB) [inline src/validation.cairo:527]: I've been trying to make this test pass but I'm not sure what I'm missing @maciejka @m-kus,

File: src/validation.cairo (hunk 2 above)
Comment (maciejka) [reply src/validation.cairo:221]: Don't worry about utreexo for now. It will be done as a separate issue.

File: src/validation.cairo (hunk 8 above)
Comment (maciejka) [reply src/validation.cairo:527]: There are compilation errors:
image
you need to mock witness with `@""` to fix this one.

image
add `?` at the end of the validate_coinbase function

File: src/validation.cairo [hunk 9]
```diff
@@ -193,8 +196,40 @@ fn fee_and_merkle_root(block: @Block) -> Result<(u256, u256), ByteArray> {
...
+
+    // Validate the outputs' amounts
+    // Sum up the total amount of all outputs
+    // and also add the outputs to the UtreexoSet.
+    let mut total_output_amount = 0;
+
+    for txs in *block
+        .txs {
+            for output in *txs.outputs {
+                total_output_amount + (*output.value);
```
Comment (maciejka) [inline src/validation.cairo:220]: `total_output_amount += *output.value`?

File: src/validation.cairo [hunk 10]
```diff
@@ -350,4 +386,182 @@ mod tests {
...
+                        .span(),
+                    lock_time: 0
+                }
+            ]
+                .span()
+        };
+        let total_fees = 5000000000_u64;
+        let block_height = 1;
+
+        validate_coinbase(@block, total_fees.into(), block_height);
```
Comment (maciejka) [inline src/validation.cairo:439]: add "unwrap()" at the end

File: src/validation.cairo [hunk 11]
```diff
@@ -350,4 +386,182 @@ mod tests {
...
```
Comment (maciejka) [inline src/validation.cairo:481]: add "unwrap()" at the end

File: src/validation.cairo [hunk 12]
```diff
@@ -350,4 +386,182 @@ mod tests {
...
```
Comment (maciejka) [inline src/validation.cairo:523]: add "unwrap()" at the end

**Review by maciejka — CHANGES_REQUESTED** [review]

File: src/validation.cairo (hunk 8 above)
Comment (manlikeHB) [reply src/validation.cairo:527]: I'm not getting the error on "witness" locally, all the other 3 tests are passing except the last one (test_validate_coinbase_outputs_amount)
Screenshot 2024-08-13 at 9 38 43 AM

File: src/validation.cairo (hunk 9 above)
Comment (manlikeHB) [reply src/validation.cairo:220]: Oh! thanks.

File: src/validation.cairo (hunk 8 above)
Comment (maciejka) [reply src/validation.cairo:527]: did you push your recent changes?

Comment (manlikeHB) [reply src/validation.cairo]: That screen shot was from before I made the requested changes.

**Pull request merged**